### PR TITLE
Type check for private constants

### PR DESCRIPTION
**Pull request by bss**
This adds support for type checking private constants statically:
```ruby
class Foo
  BAR = 123
  private_constant :BAR
end
Foo::BAR # error: private constant Foo::BAR referenced
```

### Motivation
This is currently not checked statically by sorbet.

### Test plan
See included automated tests.

### CODE REVIEW

File: namer/namer.cc
Comment (jez) [inline namer/namer.cc:41]: Why are we adding a new DefinitionKind?

File: resolver/resolver.cc
Comment (jez) [inline resolver/resolver.cc:231]: What is a local lookup?

> Any resolved constant that is private should by definition error our.

What do you mean by this? I think you mean to mention something about resolving a constant that had a `c.scope` set, specifically.

```suggestion
            // We're relying on the fact that any local lookups of the constant would already have happened.
            // Any resolved constant that is private should by definition error out.
```

File: namer/namer.cc
Comment (jez) [reply namer/namer.cc:41]: cc @jvilk if you have a second to look at this and see if this is how you'd implement private constant support or if you'd re-use one of the existing `DefinitionKind`s, that'd be great.

Comment (bss) [reply namer/namer.cc:41]: @jez @jvilk 
`private_constant` can be called for both a class/module as well as a static field. When handling the `private_constant` call in the `SymbolFinder`, we don't yet know if the argument will refer to a `Class` or a `StaticField`. We don't know that until the `SymbolDefiner` pass. I could maybe repurpose one of the other types in the enum, but I think that would make it harder to reason about.

An alternative could be to stop using `DefinitionKind` in the `Modifier` struct and instead create a new enum for this, it could look something like this:
```
enum class ModifierKind : u1 {
    Class = 0,
    Method = 1,
    ClassOrStaticField = 2,
};
```

Open to suggestions here.

File: resolver/resolver.cc
Comment (bss) [reply resolver/resolver.cc]: You are absolutely right, I don't know what I was thinking. I've updated the message to something that's hopefully more meaningful.

File: test/testdata/infer/private_constant.rb
Comment (jez) [inline test/testdata/infer/private_constant.rb:48]: Can you add tests for

- private type aliases (type aliases are `static-field` symbols that have a special bit set on them to make them type aliases, so they should work without further modification but I'd like to test
- usages of private classes / modules / type aliases / class aliases in signatures

this is a class alias:

```ruby
# typed: true
A = Integer
private_constant :A

extend T::Sig
sig {params(x: A).void}
def foo(x); end
```

Can you also add a test for type member symbols?

```ruby
# typed: true
extend T::Sig

class A
  extend T::Generic
  Elem = type_member

  class B
    extend T::Sig
    sig {params(x: Elem).void}
    def bar(x); end
  end
end

sig {params(x: A::Elem).void}
def foo(x); end
```
[→ View on sorbet.run](https://sorbet.run/#%23%20typed%3A%20true%0Aextend%20T%3A%3ASig%0A%0Aclass%20A%0A%20%20extend%20T%3A%3AGeneric%0A%20%20Elem%20%3D%20type_member%0A%0A%20%20class%20B%0A%20%20%20%20extend%20T%3A%3ASig%0A%20%20%20%20sig%20%7Bparams(x%3A%20Elem).void%7D%0A%20%20%20%20def%20bar(x)%3B%20end%0A%20%20end%0Aend%0A%0Asig%20%7Bparams(x%3A%20A%3A%3AElem).void%7D%0Adef%20foo(x)%3B%20end)

We already have some ad-hoc logic to make sure that you don't refer to type members outside of their class, because the rules that we want to implement are actually stricter than `private_constant` (see the example above—`bar`'s sig has a private reference to the constant, but that's not valid because `Elem` wasn't defined in `B`, but in `A`).

Given that we must keep the special handling of type members, I'd suggest that you implement straightforward support for private type members the same way as static fields. Even though our checks for proper type member usage are stricter than private checks, i don't think we attempt to build those type member checks on top of private checks, because those constants will not actually be marked private at runtime.

File: namer/namer.cc
Comment (jez) [reply namer/namer.cc:40]: @bss after looking at the code a bit, I think your suggestion to introduce a new `ModifierKind` enum is a great idea. That gets at exactly the root of the problem I had with adding a new definition kind—we should never find new "constant" definitions, but it's totally fine for us to modify "any existing definition which is a constant".

If you want to implement that I'd be happy taking this change!

Comment (jez) [reply namer/namer.cc:40]: fwiw, i would nest that new enum class inside `struct Modifier` so that it's `Modifier::Kind`.

File: namer/namer.cc
Comment (jez) [inline namer/namer.cc:1359]: Given the suggestion above of changing to `Modifier::Kind` will let us delete this `default` case so that clang can check for exhaustiveness here.

File: namer/namer.cc
Comment (jez) [inline namer/namer.cc:133]:
```suggestion
    enum class Kind : u1 {
```

https://stackoverflow.com/a/32953854

`enum class` is better in almost every way, because `enum` is not much better than an integer.

File: namer/namer.cc
Comment (jez) [inline namer/namer.cc:137]:
```suggestion
    };
    Kind kind;
```

i prefer no anonymous enums, so that you can write them in type signatures if you need to.

File: namer/namer.cc
Comment (jez) [reply namer/namer.cc:1362]: Bump on this. If you use `enum class`, you should be able to delete the `default` case and clang will check  you for exhaustiveness.

File: test/testdata/infer/private_constant.rb
Comment (jez) [inline test/testdata/infer/private_constant.rb:99]: Do we not additionally report a private constant access type error here? I would have thought that we'd print both but maybe we early exit after printing the first error or something.

File: namer/namer.cc
Comment (jez) [inline namer/namer.cc:1048]:
```suggestion
        if (constant.exists() && mod.name == core::Names::privateConstant()) {
```

Comment (jez) [reply namer/namer.cc:1048]: technically `_id` is a "private" field but we don't actually make it private because otherwise there's unfortunately no way to `switch` over NameRef's. So we only use it in `switch` / `case` statements.

File: namer/namer.cc
Comment (jez) [inline namer/namer.cc:1052]:
```suggestion
            } else if (constant.data(ctx)->isStaticField()) {
                constant.data(ctx)->setStaticFieldPrivate();
            } else if (constant.data(ctx)->isTypeMember()) {
                // Visibility on type members is special (even more restrictive than private),
                // so we ignore requests to mark type members private.
```

Comment (jez) [reply namer/namer.cc:1052]: no op branch for type members to be self documenting

**Review by jez — APPROVED** [review]
This looks very good. I have some suggestions that I would like, but once those are fixed I'm happy to merge.

**Comment by jez** [comment]
We have a policy of testing changes to Sorbet against Stripe's codebase before
merging them. I've kicked off a test run for the current PR.  When the build
finishes, I'll share with you whether or how it failed. Thanks!

Stripe employees can see the build result here:

→ https://go/builds/bui_IPUUmwR9hgPYKR

File: resolver/resolver.cc
Comment (jez) [inline resolver/resolver.cc:249]:
```suggestion
                    e.setHeader("Non-private reference to private constant `{}` referenced", result.data(ctx)->show(ctx));
```

- Makes this error message use the same language as for private methods
- the backticks around the `{}` will turn the interpolated string cyan

File: test/testdata/infer/private_constant.rb
Comment (bss) [reply test/testdata/infer/private_constant.rb:99]: It exits early I believe - I think that's ok though.

Comment (jez) [reply test/testdata/infer/private_constant.rb:99]: Ah sorry this comment was from part of the review where i didn't realize that we were ignoring the `modifyConstant` case for type member symbols. It makes sense that we don't print this now.

**Comment by jez** [comment]
Great change, thanks!

**Comment by reitermarkus** [comment]
This introduced a regression. We have the following in `hidden-definitions`:

```
class UnpackStrategy::Zip
  include ::UnpackStrategy::Zip::MacOSZipExtension
end
```

where `UnpackStrategy::Zip::MacOSZipExtension` is a private constant. This now fails.

**Comment by bss** [comment]
@reitermarkus was that generated using `srb rbi hidden-definitions` ? We can tweak that generation to generate code of the form:
```ruby
module UnpackStrategy
  class zip
    include MacOSZipExtension
  end
end
```

If it's a manual definition or generated using another script it would need to use that form to be able to utilize the same private constants in the rbi as the actual code is able to.

**Comment by reitermarkus** [comment]
Yes, it's generated by `srb rbi hidden-definitions`.

**Comment by bss** [comment]
@reitermarkus I see the issue here in the way hidden definitions are generated. Thanks for the bug report. I'll work on a patch for that.

In the meantime, you have a few options to unblock yourself:
- Add the private constant error code to the deny list in your sorbet config: `--error-black-list 5061`.
- Manually patch the hidden definitions file to the style I suggested above and wait with re-generating hidden-definitions until the root cause is fixed.

**Pull request merged**
